File: tests/test_pty.c
#include <stdio.h>
//#include <pty.h>
//#include <utmp.h>
#include <sys/types.h>
#include <sys/ioctl.h>
#include <termios.h>
#include <libutil.h>

int main(void)
{
    int fdm, fds, n;
    char name[1024];

    openpty(&fdm, &fds, name, 0, 0);
    printf("tty: %s\n", name);

    if (fork() == 0) {
        close(fdm);
        close(1);
        dup(fds);
        close(2);
        dup(fds);
        execlp("python", "python", "/home/huanghao/workspace/catapult/tests/echo.py", 0);
    } else {
        close(fds);
        while ((n = read(fdm, name, sizeof(name))) > 0)
        {
            printf("\n>>%.*s", n, name);
        }
    }
    return 0;
}

File: tests/test_pty.py
import os
import pty
import select
import signal
import time

master, slave = pty.openpty()

pid = os.fork()
if pid == 0:
    os.close(master)
    os.dup2(slave, 0)
    os.dup2(slave, 1)
    os.dup2(slave, 2)
    if slave > 2:
        os.close(slave)
    os.execlp('python', 'python', './echo.py')
else:
    os.close(slave)
    os.write(master, "abc")
    while 1:
        msg = os.read(master, 1024)
        if not msg:
            break
        print '<<', msg, '>>'
        time.sleep(1)
        os.kill(pid, signal.SIGINT)
    os.waitpid(pid, 0)
    print 'end'
